README.md
# RCP Webstarter Project
Ein Webstarter für RCP Anwendungen die irgendwo im Netz als zip File abliegen und keinen laufenden Servlet Container zum Download benötigen.

# Status
[![Download](https://api.bintray.com/packages/funthomas424242/eclipse-features/rcpwebstarter/images/download.svg)](https://bintray.com/funthomas424242/eclipse-features/rcpwebstarter/_latestVersion)

Letzter Shapshot:
[![Build Status](https://travis-ci.org/FunThomas424242/rcpwebstarter.svg?branch=master)](https://travis-ci.org/FunThomas424242/rcpwebstarter)

Das Projekt basiert auf den Arbeiten folgender Projekte:
* http://sourceforge.net/projects/webrcp/
* https://github.com/w11k/webrcp
* https://github.com/FunThomas424242/webrcp.example


# Benutzung
Der Webstarter ist selbst signiert. Das hat zur Folge, dass zum einen das Zertifikat in 3 Monaten ablaufen wird als auch das man für Webstart
eine Ausnahme für den Download URL aufnehmen muss. Beides ist unschön aber da ich nur ein einfacher Open Source Entwickler bin und keine Firma
fällt es mir schwer ein Zertifikat zu bekommen (Kostet vermutlich Geld und gilt vermutlich nur für einen URL und letzterer ändert sich bei mir recht
häufig).

Wer ein eigenes Zertifikat besitzt kann die sourcen downloaden und entsprechend seiner Gegebenheiten konfigurieren (build.properties, build.filter).
Alle Anderen verfahren wie hier beschrieben.

Wie auch immer, ein Test wird nur funktionieren wenn eine Ausnahme im webstart control auf dem Reiter Sicherheit eingefügt wird. Das Webstart Control
bitte auf der Kommandozeile mittels *jcontrol* starten und da wie im nachfolgenden Bild gezeigt die Ausnahme hinzufügen.
![Add Security Exception](src/main/docs/AddSecurityException.png)

# Test 
Um das Webstarter Beispiel zu testen kann nach Aufnahme der Security Ausnahme einfach auf die
[rcpwebstarter.jnlp](https://bintray.com/artifact/download/funthomas424242/eclipse-features/rcpwebstarter/1.0.0/rcpwebstarter.jnlp)
geklickt werden.

# Konfiguration für die eigene RCP Anwendung

Alles was zu tun ist sind 3 Schritte:

1. RCP Anwendung erstellen und zum Zip Archiv zusammen packen und irgendwo im Netz ablegen.
   Beim Export der RCP Anwendung aus Eclipse muss das root Verzeichnis *eclipse* genannt werden.

2. Die JNLP Datei vom Starter downloaden und den Wert des Property *jnlp.WebRCP.archives* auf den URL zum Zip Archiv der RCP Anwendung setzen.

3. RCP Zip Archiv und angepasste JNLP Datei irgendwo zusammen im Netz ablegen. 

![JNLP Datei Source](rcpwebstarter.jnlp)


CHANGES
1.0.0-SNAPSHOT 
==============
* initial version


src/main/java/com/w11k/webrcp/UnpackThread.java
/*******************************************************************************
 * Copyright (c) 2013 WeigleWilczek GmbH formerly iMedic GmbH
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.opensource.org/licenses/eclipse-1.0.php
 *
 * Contributors:
 *   WeigleWilczek GmbH [http://www.w11k.com] - initial API and implementation
 *******************************************************************************/

package com.w11k.webrcp;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Enumeration;
import java.util.List;
import java.util.Vector;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

import javax.swing.JOptionPane;
import javax.swing.SwingUtilities;

/**
 * Thread which unpacks files. The unpacking is moved in a seperate thread
 * because it can be done concurrently with the downloads.
 * 
 * @author Daniel Mendler <mendler@imedic.de>
 */
public class UnpackThread extends Thread {
	private List<File> fileList;

	private File destDir;

	private boolean finished, cleanUp;

	/**
	 * Constructor
	 */
	public UnpackThread(File destDir, boolean cleanUp) {
		this.destDir = destDir;
		this.cleanUp = cleanUp;
		fileList = new Vector<File>();
		finished = false;
		start();
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Runnable#run()
	 */
	public void run() {
		if (cleanUp)
			cleanUpDestDir();
		while (!finished || !fileList.isEmpty()) {
			if (fileList.isEmpty()) {
				synchronized (this) {
					try {
						wait();
					} catch (InterruptedException ex) {
					}
				}
			}

			// Empty after notification -> terminate
			if (fileList.isEmpty())
				return;

			unpackFile((File) fileList.remove(0));
		}
	}

	/**
	 * Add file to list of files which need unpacking
	 */
	public void addNextFile(File file) {
		fileList.add(file);
		synchronized (this) {
			notifyAll();
		}
	}

	/**
	 * Wait for thread to finish
	 */
	public void finish() {
		finished = true;
		synchronized (this) {
			notifyAll();
		}

		try {
			join();
		} catch (InterruptedException ex) {
		}
	}

	/*
	 * Unpack single file
	 */
	private void unpackFile(final File file) {
		try {
			ZipFile zipFile = new ZipFile(file);
			
			Enumeration entries = zipFile.entries();
			while (entries.hasMoreElements()) {
				ZipEntry entry = (ZipEntry) entries.nextElement();

				File entryFile = new File(destDir, entry.getName());
				if (entryFile.exists())
					continue;

				if (entry.isDirectory()) {
					entryFile.mkdirs();
					if (!entryFile.exists() || !entryFile.isDirectory())
						throw new IOException("Couldn't create directory");
				} else {
					InputStream in = zipFile.getInputStream(entry);
					OutputStream out = new FileOutputStream(entryFile);
					byte[] buffer = new byte[32768];
					int size;
					while ((size = in.read(buffer)) > 0)
						out.write(buffer, 0, size);
					in.close();
					out.close();
				}
			}

			zipFile.close();
		} catch (final IOException ex) {
			try {
				SwingUtilities.invokeAndWait(new Runnable() {
					public void run() {
						JOptionPane.showMessageDialog(null,
								"Temporary Zip-File " + file
										+ " couldn't be extracted: " + ex,
								"Extraction Error", JOptionPane.ERROR_MESSAGE);
					}
				});
			} catch (Exception ex2) {
			}
			System.exit(1);
		}
	}

	/*
	 * Clean up destination directory for unpacking
	 */
	private void cleanUpDestDir() {
		if (destDir.exists()) {
			File[] file = destDir.listFiles();
			for (int i = 0; i < file.length; ++i) {
				// Don't delete the workspace directory
				// (It contains the configuration)!
				if (!file[i].getName().equals("workspace"))
					removeRecursively(file[i]);
			}
		}
	}

	/*
	 * Remove directory structure recursively
	 */
	private void removeRecursively(File dir) {
		if (!dir.exists())
			return;

		if (!dir.isDirectory())
			dir.delete();

		File[] file = dir.listFiles();
		if (file != null) {
			for (int i = 0; i < file.length; ++i)
				removeRecursively(file[i]);
		}

		dir.delete();
	}
}

src/main/java/com/w11k/webrcp/AllPermissionPolicy.java
/*******************************************************************************
 * Copyright (c) 2013 WeigleWilczek GmbH formerly iMedic GmbH
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.opensource.org/licenses/eclipse-1.0.php
 *
 * Contributors:
 *   WeigleWilczek GmbH [http://www.w11k.com] - initial API and implementation
 *******************************************************************************/

package com.w11k.webrcp;

import java.security.AllPermission;
import java.security.CodeSource;
import java.security.PermissionCollection;
import java.security.Permissions;
import java.security.Policy;
import java.security.ProtectionDomain;

/**
 * Permission policy which allows everything to every codebase and domain. This
 * permission will be loaded because the manually loaded classes (e.g. the
 * eclipse launcher) haven't enough permissions with the standard web start
 * policy.
 * 
 * @author Daniel Mendler <mendler@imedic.de>
 */
public class AllPermissionPolicy extends Policy {
	public PermissionCollection getPermissions(CodeSource code) {
		Permissions permissions = new Permissions();
		permissions.add(new AllPermission());
		return permissions;
	}

	public PermissionCollection getPermissions(ProtectionDomain domain) {
		Permissions permissions = new Permissions();
		permissions.add(new AllPermission());
		return permissions;
	}

	public void refresh() {
	}
}

src/main/java/com/gh/funthomas424242s/rcpwebstarter/Logger.java
package com.gh.funthomas424242s.rcpwebstarter;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.JOptionPane;
import javax.swing.ProgressMonitor;
import javax.swing.Timer;

public class Logger<T> {

	final static boolean DEBUG_MODE=false;
	
	final Class<T> callerClass;

	public Logger(final Class<T> callerClass) {
		this.callerClass = callerClass;
	}

	public ProgressMonitor createProgressMonitor(String message, int max) {
		ProgressMonitor pm = new ProgressMonitor(null, message, "", 0, max);
		pm.setMillisToDecideToPopup(100);
		pm.setMillisToPopup(500);

		return pm;
	}

	/*
	 * Simulate progress by showing a faked progress monitor
	 */
	public void simulateProgress(String text, int seconds) {
		final ProgressMonitor pm = this.createProgressMonitor(text,
				seconds * 1000 / 50);

		Timer timer = new Timer(50, new ActionListener() {
			private int progress = 0;

			public void actionPerformed(ActionEvent event) {
				if (pm.isCanceled()) {
					System.exit(0);
				}
				pm.setProgress(progress += 1);
			}
		});

		timer.start();
	}

	public void printDebugInfoBox(String title, String message) {
		if (DEBUG_MODE) {
			JOptionPane.showMessageDialog(null, message, title,
					JOptionPane.ERROR_MESSAGE);
		}
	}

	public void beendeMitFehler(String title, String message) {
		JOptionPane.showMessageDialog(null, message, title,
				JOptionPane.ERROR_MESSAGE);
		System.exit(1);
	}

}


src/main/java/com/gh/funthomas424242s/rcpwebstarter/Downloader.java
package com.gh.funthomas424242s.rcpwebstarter;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URL;
import java.net.URLConnection;

import javax.jnlp.BasicService;
import javax.jnlp.DownloadService;
import javax.swing.ProgressMonitor;

import com.gh.funthomas424242s.rcpwebstarter.env.WebStartUmgebung;

public class Downloader {
	
	final static Logger<Downloader> LOGGER = new Logger<Downloader>(Downloader.class);
	final protected BasicService basicService;
	final protected DownloadService downloadService;
	
	public Downloader(final WebStartUmgebung webstartUmgebung){
		this.basicService=webstartUmgebung.getBasicService();
		this.downloadService=webstartUmgebung.getDownloadService();
	}
	
	
	/*
	 * Download a file from an url and store it at destFile. A progress monitor
	 * will be shown.
	 */
	protected void downloadFile(URL url, File destFile) {
		try {

			destFile.getParentFile().mkdirs();
			LOGGER.printDebugInfoBox("DestFile2 ", destFile.getAbsolutePath());
			OutputStream out = new FileOutputStream(destFile);

			URLConnection conn = url.openConnection();
			InputStream in = conn.getInputStream();

			int totalSize = conn.getContentLength(), downloadedSize = 0, size;
			byte[] buffer = new byte[32768];

			ProgressMonitor pm = LOGGER.createProgressMonitor("Downloading " + url,
					totalSize);
			boolean canceled = false;

			while ((size = in.read(buffer)) > 0
					&& !(canceled = pm.isCanceled())) {
				out.write(buffer, 0, size);
				pm.setProgress(downloadedSize += size);
				// pm.setNote(downloadedSize / totalSize + "% finished");
			}

			in.close();
			out.close();

			if (canceled) {
				destFile.delete(); // Delete uncomplete file
				LOGGER.beendeMitFehler("Starting canceled",
						"Downloading canceled. Exiting...");
			}

			pm.close();
		} catch (IOException ex) {
			final StringBuilder builder = new StringBuilder();
			final StackTraceElement[] arr = ex.getStackTrace();
			for (StackTraceElement s : arr) {
				builder.append(s.toString());
			}
			LOGGER.beendeMitFehler("Download Error", "Couldn't download file: "
					+ builder.toString());
		}
	}
	
	

	
	

}


src/main/java/com/w11k/webrcp/WebRCP.java
/*******************************************************************************
 * Copyright (c) 2013 WeigleWilczek GmbH formerly iMedic GmbH
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.opensource.org/licenses/eclipse-1.0.php
 *
 * Contributors:
 *   WeigleWilczek GmbH [http://www.w11k.com] - initial API and implementation
 *******************************************************************************/

package com.w11k.webrcp;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.io.Writer;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.MalformedURLException;
import java.net.ServerSocket;
import java.net.URL;
import java.net.URLClassLoader;
import java.net.URLConnection;
import java.security.Policy;
import java.util.Enumeration;
import java.util.Locale;
import java.util.Properties;
import java.util.Set;

import javax.jnlp.BasicService;
import javax.jnlp.ServiceManager;
import javax.jnlp.UnavailableServiceException;
import javax.swing.JOptionPane;
import javax.swing.ProgressMonitor;
import javax.swing.Timer;

import sun.reflect.ReflectionFactory.GetReflectionFactoryAction;

/**
 * WebRCP - Web Start Application which acts as loader for an Eclipse RCP
 * application.
 * 
 * @author by Daniel Mendler <mendler@imedic.de>
 */
public class WebRCP {
	/*
	 * Supported system architectures (From org.eclipse.core.runtime.Platform)
	 */
	private static final String ARCH_X86 = "x86";
	private static final String ARCH_PA_RISC = "PA_RISC";
	private static final String ARCH_PPC = "ppc";
	private static final String ARCH_SPARC = "sparc";
	private static final String ARCH_AMD64 = "amd64";
	private static final String ARCH_X86_64 = "x86_64";
	private static final String ARCH_IA64 = "ia64";

	/*
	 * Supported windowing systems (From org.eclipse.core.runtime.Platform)
	 */
	private static final String WS_WIN32 = "win32";
	private static final String WS_MOTIF = "motif";
	private static final String WS_GTK = "gtk";
	private static final String WS_PHOTON = "photon";
	private static final String WS_CARBON = "carbon";

	/*
	 * Supported operating systems (From org.eclipse.core.runtime.Platform)
	 */
	private static final String OS_WIN32 = "win32";
	private static final String OS_LINUX = "linux";
	private static final String OS_AIX = "aix";
	private static final String OS_SOLARIS = "solaris";
	private static final String OS_HPUX = "hpux";
	private static final String OS_QNX = "qnx";
	private static final String OS_MACOSX = "macosx";

	/*
	 * Configuration property names
	 */
	private static final String PROPERTY_BASEURL = "jnlp.WebRCP.baseURL";
	private static final String PROPERTY_APPNAME = "jnlp.WebRCP.appName";
	private static final String PROPERTY_APPVERSION = "jnlp.WebRCP.appVersion";
	private static final String PROPERTY_LAUNCHAPP = "jnlp.WebRCP.launchApp";
	private static final String PROPERTY_LAUNCHPRODUCT = "jnlp.WebRCP.launchProduct";
	private static final String PROPERTY_ARCHIVES = "jnlp.WebRCP.archives";
	private static final String PROPERTY_SINGLEINST = "jnlp.WebRCP.singleInstance";
	private static final String PROPERTY_EXECUTABLE = "jnlp.WebRCP.executable";

	/*
	 * Port used to check for a running instance. This port should be hopefully
	 * unused.
	 */
	private static final int SINGLEINST_PORT = 25975;

	/*
	 * Eclipse launcher constants
	 */
	private static final String LAUNCHER_CLASS = "jnlp.WebRCP.launcherclass";
	private static final String LAUNCHER_JAR = "jnlp.WebRCP.launcherjar";

	/*
	 * Try to determine system architecture by examining the system property
	 * "os.arch"
	 */
	private static String determineArch() {
		String arch = System.getProperty("os.arch").toLowerCase();

		printInfoBox("Arch",arch);
		
		if (arch.indexOf("x86") >= 0 || arch.matches("i.86"))
			return ARCH_X86;

		if (arch.indexOf("ppc") >= 0 || arch.indexOf("power") >= 0)
			return ARCH_PPC;

		if (arch.indexOf("x86_64") >= 0 || arch.indexOf("amd64") >= 0)
			return ARCH_AMD64;

		if (arch.indexOf("ia64") >= 0)
			return ARCH_IA64;

		if (arch.indexOf("risc") >= 0)
			return ARCH_PA_RISC;

		if (arch.indexOf("sparc") >= 0)
			return ARCH_SPARC;

		handleError("Unknown Architecture",
				"Your system has an unknown architecture: " + arch);

		return null;
	}

	/*
	 * Try to determine operating system by examining the system property
	 * "os.name"
	 */
	private static String determineOS() {
		String os = System.getProperty("os.name").toLowerCase();

		if (os.indexOf("linux") >= 0)
			return OS_LINUX;

		if (os.indexOf("mac") >= 0)
			return OS_MACOSX;

		if (os.indexOf("windows") >= 0)
			return OS_WIN32;

		if (os.indexOf("hp") >= 0 && os.indexOf("ux") >= 0)
			return OS_HPUX;

		if (os.indexOf("solaris") >= 0)
			return OS_SOLARIS;

		if (os.indexOf("aix") >= 0)
			return OS_AIX;

		if (os.indexOf("qnx") >= 0)
			return OS_QNX;

		handleError("Unknown Operating System",
				"Your operating system is unknown: " + os);

		return null;
	}

	/*
	 * Get window system by operating system name
	 */
	private static String getWindowSystem(String os) {
		if (os.equals(OS_WIN32))
			return WS_WIN32;

		if (os.equals(OS_LINUX))
			return WS_GTK;

		if (os.equals(OS_QNX))
			return WS_PHOTON;

		if (os.equals(OS_MACOSX))
			return WS_CARBON;

		return WS_MOTIF; // OS_AIX, OS_HPUX, OS_SOLARIS
	}

	/*
	 * Ensure that there's only one application instance running.
	 */
	private static void ensureSingleInstance() {
		try {
			new ServerSocket(SINGLEINST_PORT);
		} catch (Exception ex) {
			handleError("Already running.",
					"There's already an instance running.");
		}
	}

	/*
	 * Get base URL
	 */
	private static String getBaseURL() {
		try {
			BasicService service = (BasicService) ServiceManager
					.lookup("javax.jnlp.BasicService");
			return service.getCodeBase().toString();
		} catch (UnavailableServiceException ex) {
			handleError("WebStart Service Error",
					"Service javax.jnlp.BasicService unvailable: " + ex);
			return null;
		}
	}

	/*
	 * Check for new eclipse application version by comparing the value in the
	 * version file and the version argument. The version value can be an
	 * arbitrary string.
	 */
	private static boolean newVersionAvailable(String newVersion,
			File versionFile) {
		String oldVersion = null;

		try {
			// Read old version
			BufferedReader in = new BufferedReader(new FileReader(versionFile));
			oldVersion = in.readLine();
			in.close();
		} catch (IOException ex) {
			// No error. File doesn't already exists.
		}

		try {
			// Write new version
			Writer out = new FileWriter(versionFile);
			out.write(newVersion);
			out.close();

			return oldVersion == null || !newVersion.equals(oldVersion);
		} catch (IOException ex) {
			// Not too bad. We continue.
			return true;
		}
	}

	/*
	 * Download a file from an url and store it at destFile. A progress monitor
	 * will be shown.
	 */
	private static void downloadFile(URL url, File destFile) {
		try {
			OutputStream out = new FileOutputStream(destFile);

			URLConnection conn = url.openConnection();
			InputStream in = conn.getInputStream();

			int totalSize = conn.getContentLength(), downloadedSize = 0, size;
			byte[] buffer = new byte[32768];

			ProgressMonitor pm = createProgressMonitor("Downloading " + url,
					totalSize);
			boolean canceled = false;

			while ((size = in.read(buffer)) > 0
					&& !(canceled = pm.isCanceled())) {
				out.write(buffer, 0, size);
				pm.setProgress(downloadedSize += size);
				// pm.setNote(downloadedSize / totalSize + "% finished");
			}

			in.close();
			out.close();

			if (canceled) {
				destFile.delete(); // Delete uncomplete file
				handleError("Starting canceled",
						"Downloading canceled. Exiting...");
			}

			pm.close();
		} catch (IOException ex) {
			final StringBuilder builder = new StringBuilder();
			final StackTraceElement[] arr = ex.getStackTrace();
			for (StackTraceElement s : arr) {
				builder.append(s.toString());
			}
			handleError("Download Error",
					"Couldn't download file: " + builder.toString());
		}
	}

	/*
	 * Load and start eclipse launcher org.eclipse.core.launcher.Main
	 */
	private static void startLauncher(String url, String os, String arch,
			String arg) {
		URLClassLoader urlClassLoader = null;
		try {
			// Reload new policy which allows all to all codebases
			// because the default policy doesn't apply to the code loaded from
			// startup.jar!!!
			Policy.setPolicy(new AllPermissionPolicy());

			final String path = "file:" + url + File.separator
					+ getSystemProperty(LAUNCHER_JAR);
			System.out.println("launcher jar: " + path);
			final String launcherClassName = getSystemProperty(LAUNCHER_CLASS);
			// TODO
			printInfoBox("Launcher Jar", path);
			printInfoBox("Launcher Class", launcherClassName);

			urlClassLoader = new URLClassLoader(new URL[] { new URL(path) });
			Class<?> launcher = urlClassLoader.loadClass(launcherClassName);
			final Method launcherMain = launcher.getMethod("main", String.class);
			
			/*
			 * Start launcher with aurguments -os <operating-system> -ws
			 * <window-system> -arch <system architecture> -install
			 * <installation directory> -data <workspace directory> -user
			 * <workspace directory> -nl <locale> <-application|product name>
			 * The default workspace directory is put under the installation
			 * directory.
			 */
			final String exeCommand = "-os " + os + " -ws "
					+ getWindowSystem(os) + " -arch " + arch + " -install "
					+ url +"/eclipse/" + " -data " + url + "/workspace/ -user " + url
					+ "/workspace/ -nl " + Locale.getDefault() + arg;

			printInfoBox("Launcher Command", exeCommand == null ? "null"
					: exeCommand);

			launcherMain.invoke(null,  exeCommand );
			
			//java -cp startup.jar org.eclipse.core.launcher.Main -os linux -ws gtk -arch amd64 -install /tmp/WebRCP Tutorial/unpacked -data /tmp/WebRCP Tutorial/unpacked/workspace/ -user /tmp/WebRCP Tutorial/unpacked/workspace/ -nl de_DE -product org.eclipse.ui.tutorials.rcp.part3.RcpApplication

			
			
		} catch (InvocationTargetException ex) {
			handleError("Startup Error", "Invocation failed: " + ex.getCause());
		} catch (IllegalAccessException ex) {
			handleError("Startup Error", "Invocation failed: " + ex);
		} catch (NoSuchMethodException ex) {
			ex.printStackTrace();
			handleError("Startup Error", "Invalid Eclipse Launcher: " + ex);
		} catch (ClassNotFoundException ex) {
			ex.printStackTrace();
			handleError("Startup Error", "Eclipse Launcher not found: " + ex);
		} catch (MalformedURLException ex) {
			// This shouldn't happen.
			throw new RuntimeException(ex);
		} finally {
			if (urlClassLoader != null) {
				try {
					urlClassLoader.close();
				} catch (IOException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
			}
		}
	}

	/*
	 * Convenience method to create a progress monitor
	 */
	private static ProgressMonitor createProgressMonitor(String message, int max) {
		ProgressMonitor pm = new ProgressMonitor(null, message, "", 0, max);
		pm.setMillisToDecideToPopup(100);
		pm.setMillisToPopup(500);

		return pm;
	}

	/*
	 * Convenience method to show an error dialog
	 */
	private static void handleError(String title, String message) {
		JOptionPane.showMessageDialog(null, message, title,
				JOptionPane.ERROR_MESSAGE);
		System.exit(1);
	}

	private static void printInfoBox(String title, String message) {
		JOptionPane.showMessageDialog(null, message, title,
				JOptionPane.ERROR_MESSAGE);
	}

	/*
	 * Simulate progress by showing a faked progress monitor
	 */
	private static void simulateProgress(String text, int seconds) {
		final ProgressMonitor pm = createProgressMonitor(text,
				seconds * 1000 / 50);

		Timer timer = new Timer(50, new ActionListener() {
			private int progress = 0;

			public void actionPerformed(ActionEvent event) {
				if (pm.isCanceled()) {
					System.exit(0);
				}
				pm.setProgress(progress += 1);
			}
		});

		timer.start();
	}

	/*
	 * Get system property or die
	 */
	private static String getSystemProperty(String key) {
		String value = System.getProperty(key);
		if (value != null)
			return value;

		handleError("Missing System Property", key + " is required");

		return null;
	}

	/**
	 * Creates a shortcut on the users desktop to the passed target
	 * 
	 * @param shortcutTarget
	 * @param appName
	 */
	private static void createDesktopShortcutToExe(String shortcutTarget,
			String appName) {
		Shortcut scut;
		try {
			scut = new Shortcut(new File(shortcutTarget));

			final String userHomeDir = System.getProperty("user.home");
			final String os = determineOS();
			String shortCutDirName = "Desktop";
			if (os.equals(OS_LINUX)) {
				shortCutDirName = "Schreibtisch";
			}
			final String shortCutFileName = userHomeDir + File.separator
					+ shortCutDirName + File.separator + appName + ".lnk";
			OutputStream outStream = new FileOutputStream(shortCutFileName);
			outStream.write(scut.getBytes());
			outStream.flush();
			outStream.close();
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
			handleError("Error while creating Desktop Shortcut",
					"Wrong Shortcut target specified" + e.getLocalizedMessage());
		} catch (FileNotFoundException e) {
			e.printStackTrace();
			handleError("Error while creating Desktop Shortcut",
					"File not found" + e.getLocalizedMessage());
		} catch (IOException e) {
			e.printStackTrace();
			handleError("Error while creating Desktop Shortcut",
					e.getLocalizedMessage());
		}
	}

	/**
	 * Run WebRCP
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		// http://stackoverflow.com/questions/19407102/java-7-update-45-broke-my-web-start-swt-application
		Properties properties = System.getProperties();
		// copy properties to avoid ConcurrentModificationException
		Properties copiedProperties = new Properties();
		copiedProperties.putAll(properties);
		Set<Object> keys = copiedProperties.keySet();
		for (Object key : keys) {
			if (key instanceof String) {
				String keyString = (String) key;
				if (keyString.startsWith("jnlp.custom.")) {
					// re set all properties starting with the
					// jnlp.custom-prefix
					// and set them without the prefix
					String property = System.getProperty(keyString);
					String replacedKeyString = keyString.replaceFirst(
							"jnlp.custom.", "");

					System.setProperty(replacedKeyString, property);
				}
			}
		}

		if (Boolean.getBoolean(PROPERTY_SINGLEINST)) {
			ensureSingleInstance();
		}

		// Get required properties
		String appName = getSystemProperty(PROPERTY_APPNAME);
		String appVersion = getSystemProperty(PROPERTY_APPVERSION);
		String[] archive = getSystemProperty(PROPERTY_ARCHIVES).split(
				"\\s*,\\s*");

		// Get application/product to launch
		String launcherArg = System.getProperty(PROPERTY_LAUNCHAPP);

		if (launcherArg == null) {
			launcherArg = System.getProperty(PROPERTY_LAUNCHPRODUCT);

			if (launcherArg == null) {
				handleError("Missing System Property", PROPERTY_LAUNCHAPP
						+ " or " + PROPERTY_LAUNCHPRODUCT + " are required");
			} else {
				launcherArg = " -product " + launcherArg;
			}
		} else {
			launcherArg = " -application " + launcherArg;
		}

		String baseURL = getBaseURL();
		String arch = determineArch();
		String os = determineOS();
		
		//Sonderfall Ubuntu
		if(arch.equals(ARCH_AMD64) && os.equals(OS_LINUX)){
			arch=ARCH_X86_64;
		}

		File tempDir = new File(getSystemProperty("java.io.tmpdir"), appName);
		tempDir.mkdirs();

		File unpackDestDir = new File(tempDir, "unpacked");

		// Check for new version
		boolean override = newVersionAvailable(appVersion, new File(tempDir,
				"version"));

		// Start background thread for unpacking
		UnpackThread unpackThread = new UnpackThread(unpackDestDir, override);

		// Download and unpack system-independant archives
		for (String element : archive) {
			File destFile = new File(tempDir, element + ".zip");

			if (!destFile.exists() || override) {

				try {
					final URL archiveURL = new URL(baseURL + "archives/"+ element + ".zip");
					// TODO
					final String message = archiveURL == null ? "null"
							: archiveURL.toString();
					printInfoBox("URL "+element, message);
					if (archiveURL != null) {
						downloadFile(archiveURL, destFile);
					}
				} catch (MalformedURLException ex) {
					// This shouldn't happen.
					throw new RuntimeException(ex);
				}
			}

			unpackThread.addNextFile(destFile);
		}

		// Show a progress monitor which "simulates" the startup of
		// the application.
		simulateProgress("Loading " + appName + "...", 3);

		// Wait for the unpacking thread to complete
		unpackThread.finish();

		// Store base url (might be used by the loaded program)
		System.setProperty(PROPERTY_BASEURL, baseURL);

		// create dekstop shortcut
		String executable = System.getProperty(PROPERTY_EXECUTABLE);

		createDesktopShortcutToExe(unpackDestDir.getPath() + File.separator
				+ executable, appName);

		// try {
		// Then start the launcher!
		// startLauncher(unpackDestDir.toURI().toURL(), os, arch, launcherArg);
		startLauncher(unpackDestDir.getPath(), os, arch, launcherArg);
		// } catch (MalformedURLException ex) {
		// // This shouldn't happen.
		// throw new RuntimeException(ex);
		// }
	}

	private static void printSystemProperties() {
		Properties p = System.getProperties();
		Enumeration keys = p.keys();
		while (keys.hasMoreElements()) {
			String key = (String) keys.nextElement();
			String value = (String) p.get(key);
			System.out.println(key + ": " + value);
		}
	}

	public static URL getResource(final String resourcePath) {
		final ClassLoader classLoader = Thread.currentThread()
				.getContextClassLoader();
		// final ClassLoader classLoader = this.getClass().getClassLoader();
		final URL resourceURL = classLoader.getResource(resourcePath);
		return resourceURL;
	}
}